Fix mod suppression when same-name trait method exists

TedDriggs opened the pull request:
Fix #680: Only show trait method suggestions when contextually relevant by checking if the cursor is in a function identifier, and otherwise not looking for trait method declarations and trait methods when searching for unprefixed values.

This is most useful when dealing with `std::fmt`; the example code for this module typically imports it at the mod level, which causes autocomplete not to work for writing the rest of the signature, where `fmt::Formatter` and `fmt::Result` will appear.

TedDriggs commented:
@jwilm I merged when I should have rebased, sorry for the noise in the PR.

jwilm commented on src/racer/nameres.rs:
```diff
@@ -1418,6 +1417,68 @@ pub fn resolve_path(path: &core::Path, filepath: &Path, pos: Point,
     }
 }
 
+pub(crate) fn resolve_method(point: Point, msrc: Src, searchstr: &str,
```
Would love to have a doc comment on this explaining what the algorithm does.

jwilm replied:
Oh, I see that part of it was just a refactor from elsewhere...

TedDriggs replied:
I added a comment explaining it a bit better.

jwilm commented on src/racer/matchers.rs:
```diff
@@ -23,7 +23,7 @@ pub type PendingImports<'stack, 'fp> = StackLinkedListNode<'stack, PendingImport
 pub type MChain<T> = iter::Chain<T, MIter>;
 
-// Should I return a boxed trait object to make this signature nicer?
+// TODO change return type to `impl Iterator<Item = Match>`
```
So excited for this.

jwilm commented on src/racer/core.rs:
```diff
@@ -955,6 +956,25 @@ fn complete_from_file_(
+            if util::in_fn_name(line) {
+                trace!("Path is in fn declaration: `{}`", expr);
+
+                for m in nameres::resolve_method(
```
The formatting on this is really hard to parse. It might be better to assign the returned Iter/IntoIter to a temporary variable first. Eg

```rust
let methods = nameres::resolve(/* ... */);
for m in methods {
    out.push(m);
}
```

Actually, maybe it would be better if resolve_method just returned a `Vec` instead of an iterator so that it may be appended directly only `out` instead of doing this iterative thing.

TedDriggs replied:
As part of #778 I changed the return type of this function to be `Vec`, which will remove the `into_iter` thing. There's honestly no need to append in this case, so I can remove the loop outright and just return the `resolve_method` return value directly.

jwilm commented on src/racer/nameres.rs:
```diff
@@ -525,6 +495,34 @@ fn search_scope_headers(point: Point, scopestart: Point, msrc: Src, searchstr: &
+/// TODO: Handle `extern` functions
+fn preblock_is_fn(preblock: &str) -> bool {
+    // Perform simple checks
+    if preblock.starts_with("fn") || preblock.starts_with("pub fn") || preblock.starts_with("const fn") {
```
Iterative approach might be cleaner here. _llvm_ should unroll the loop anyhow. Eg

```rust
for case in &["fn", "pub fn", "const fn"] {
    if preblock.starts_with(case) {
        return true;
    }
}
```

TedDriggs replied:
True; I wasn't sure at first if they were all going to be the same. I'll change this.

jwilm commented on tests/system.rs:
```diff
@@ -1285,29 +1285,37 @@ fn keeps_newlines_in_external_mod_doc() {
+/// Addresses https://github.com/racer-rust/racer/issues/618
 #[test]
-fn issue_618() {
+fn always_get_all_doc_lines() {
```
It's kind of nice to have the issue number in the test name. I like what you've done to explain what the test actually is though. Maybe we can combine them? `issue_618_always_foo`.

TedDriggs replied:
I moved the issue number to the doc comment so that I could include the full URL. I'm okay with putting issue numbers in test names for the purposes of `cargo test 618` but without easy reference to the issue I'm not sure how valuable the number is.

jwilm replied:
For me, it's more about knowing it's associated with 618 when it fails just by looking at cargo test output.

TedDriggs commented:
@jwilm I've added the tests against `std::fmt` and addressed all points on the CR except for issue IDs in test names. If it's okay with you, I'll move forward as-is and then we can put in a separate PR to add those back to the test names.

TedDriggs commented:
Git has gone beserk; closing this PR and making a new one.
